js/piwik.js

```diff
@@ -7634,6 +7645,159 @@ if (typeof window.Piwik !== 'object') {
+                };
+                script.src = getConfigsUrl;
+                // timeout may reduce the risk of delaying onload event
+                setTimeout(function () {
+                    var head = documentAlias.getElementsByTagName('head');
+                    if (head && head.length && head[0]) {
+                        head[0].appendChild(script);
+                    } else {
+                        var body = documentAlias.getElementsByTagName('body');
+                        if (body && body.length && body[0]) {
+                            body[0].appendChild(script);
+                        }
+                    }
+                }, 10);
+            }
+
+            function saveTrackerExtraConfig(idSite, config) {
+                if (typeof sessionStorage === 'undefined' || !sessionStorage) {
+                    return;
+                }
+
+                var configStr = JSON.stringify({
+                    config: config,
+                    expires: (new Date()).getTime() + (configVisitStandardLength * 60 * 1000)
+                });
+                var cacheKey = 'mtm.extraTrackerConfig.' + idSite;
+
+                sessionStorage.setItem(cacheKey, configStr);
+            }
+
+            function getCachedTrackerExtraConfig() {
+                if (typeof sessionStorage === 'undefined' || !sessionStorage) {
+                    return;
+                }
+
+                var cacheEntry;
+                var cacheKey = 'mtm.extraTrackerConfig.' + configTrackerSiteId;
+
+                try {
+                    cacheEntry = JSON.parse(sessionStorage.getItem(cacheKey));
+                } catch (e) {
+                    var consoleType = typeof console;
+                    if (consoleType !== 'undefined' && console && console.log) {
+                        console.log("Failed to parse cached extra tracker config JSON: " + (e.stack || e.message));
+                    }
+                    return;
+                }
+
+                if (!cacheEntry || !cacheEntry.expires || !cacheEntry.config) {
+                    return;
+                }
+
+                var expires = cacheEntry.expires;
+                if (expires < Date.now()) {
+                    return;
+                }
+
+                return cacheEntry.config;
+            }
+
+            /**
+             * Sets extra tracker config if the config is for this tracker's idSite. For internal use only.
+             *
+             * @param {string} trackerIdSite
+             * @param {object} config
+             * @internal
+             */
+            this.setExtraConfig = function (trackerIdSite, config) {
+                if (configTrackerSiteId != trackerIdSite
+                    || loadExtraConfigEncounteredError
+                ) {
+                    return;
+                }
+
+                saveTrackerExtraConfig(trackerIdSite, config);
```

Comment:
@diosmosis so far the PR looks straight forward. Haven't tested it yet though. 

Re saving it in session... 

How Heatmaps & Session Recording worked so far is that:
* It does a request on each page like `configs.php?url=https://example.com`
* Then the logic checks for magic heatmaps or sessions, and outputs the matching ones
* It also does some other server side logic based on configured screenshot_url etc.

Now that we cache this info (which is great), I reckon we simply return all recorded heatmap/session recording information and then do this logic client side? It means we basically do return the info of all configured heatmaps/sessions to the user but they can kind of see it anyway I suppose (if they were to view all page urls) so should be all good. 

I'm planning to spend some time next week to rewrite that logic in heatmap/session recording to verify the PR works well with plugins.

Will probably do the same for form analytics.

Reply:
👍 sounds like that makes sense, less server side logic, and just one cached query server side in the tracker, if I understand what you're saying.